### ISSUE TITLE
visual mode and adblock update

### ISSUE THREAD

**Issue by zakariaGatter** [luakit/luakit#899]
hello everyone 

i tried luakit and i love it so much, it's have everything i want 

except one thing "visual mode"

Why there is no visual mode?? and how can be added ?!

i am not a programmer but i create a simple command for luakit to update adblock list form easylist 

```lua
local add_cmds = modes.add_cmds
local actions = {
        adblock_update = {
                desc = "Update adblock from easylist official site",
                func = function (w)
                        local url = "https://easylist-downloads.adblockplus.org/easylist.txt"
                        local dir = os.getenv("HOME") .. "/.local/share/luakit/adblock"
                        downloads.add (url)
                        os.rename(downloads.default_dir .. "/easylist.txt", dir .. "/easylist.txt")
                end,
        }
}

add_cmds({{":adblock-update", actions.adblock_update },})
```

i hope this gonna be useful for someone 

thank you for your beautiful web browser

**Comment by taobert** [comment]
(I assume you're talking about vim's visual mode.)
What is it that you want from a visual mode? 
I use it to highlight areas when i want to do text transformations on discrete regions,
or to yank chunks of text. The former doesn't seem pertinent to luakit.
I _can_ see the appeal of being able to yank page text in luakit using the keyboard.

From the quick grep i just did, apart from the mouse pointer,
there doesn't really seem to be a concept of a cursor outside of text inputs,
so this would probably require implementing some sort of Caret Mode as well.
Alternatively adapting one of the chromium/firefox vimlike plugins might allow a pure-javascript solution.
While i'd probably use one of these if it were implemented, 
both sound like more effort to me than moving my hand to the mouse.
Less effort (maybe) would be adding window manager level key bindings to move the mouse cursor,
and trigger mouse down/up events, though this is unlikely to work as nicely.

Edit: This was also asked about in #183.

**Comment by zakariaGatter** [comment]
Yes i know what you mean and i didint know like this feature gonna take alot of time and effort 

all i was wantering about like u said almost all browser extantion base on vim have this mode thats why i suggested this 

thank you again for your time and for this wanderful browser, im gonna keep using luakit i love it so much

**Comment by taobert** [comment]
So as a follow up, setting `contenteditable=true` for the body (`:js document.body.setAttribute("contenteditable","true")`) allows insert mode to function on the entire page. This allows not just yanking text but also editing/deleting page text, so it's somewhat more dangerous than a visual mode, but it might provide the basis for making this work.

**Comment by taobert** [comment]
Hrmm, well. It seems there's already a `webview.enable_caret_browsing` setting which specifies "Whether keyboard navigation should be enabled", and set's the similarly named webkit property.
If this is set, then entering passthrough or insert modes will let you move the cursor around using the arrows (but not hjkl),
including highlighting sections (which automatically puts things in the primary X buffer).This also avoids the destructive nature of setting `contenteditable=true` which is probably not usually what people want.
So, a Caret Mode would be fairly simple to put together. The main annoyance with it would be the vagaries of dom element positioning, which can mean that pressing down moves to the next element, which might be positioned higher on the page.
